pkg/retainer/retainer/retainer.go

```diff
@@ -0,0 +1,299 @@
+	retainerStore := r.retainerFunc(blockID.Index())
+	_ = retainerStore.StoreBlockFailure(blockID, failureCode)
+}
+
+func (r *Retainer) RetainTransactionFailure(transactionID iotago.TransactionID, slotIndex iotago.SlotIndex, failureCode apimodels.TransactionFailureReason) {
+	retainerStore := r.retainerFunc(slotIndex)
+	_ = retainerStore.StoreTransactionFailure(transactionID, failureCode)
+}
+
+func (r *Retainer) blockStatus(blockID iotago.BlockID) (apimodels.BlockState, *model.Block, error) {
+	// we store the block on acceptance, so if block is nil and slot was already committed we know its orphaned or never existed
+	block, err := r.blockFromStorageFunc(blockID)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	// block was for sure accepted in this chain
+	if block != nil {
+		// check if finalized in this chain
+		if blockID.Index() <= r.finalizedSlotFunc() {
+			return apimodels.BlockStateFinalized, block, nil
+		}
+
+		// check if confirmed
+		if confirmed, err := r.retainerFunc(blockID.Index()).WasBlockConfirmed(blockID); err != nil {
+			return apimodels.BlockStateUnknown, nil, ierrors.Wrap(err, "failed to get block confirmed state in retainer.")
+		} else if confirmed {
+			return apimodels.BlockStateConfirmed, block, nil
+		}
+
+		return apimodels.BlockStatePending, block, nil
+	}
+
+	if blockID.Index() > r.latestCommittedSlotFunc() {
+		// orphaned (attached, but never accepted)
+		if orphaned, err := r.retainerFunc(blockID.Index()).WasBlockOrphaned(blockID); err != nil {
+			return apimodels.BlockStateUnknown, nil, ierrors.Wrap(err, "failed to get block orphaned state in retainer.")
+		} else if orphaned {
+			return apimodels.BlockStateOrphaned, nil, nil
+		}
+	}
+
+	if block := r.blockFromCacheFunc(blockID); block != nil {
+		return apimodels.BlockStatePending, block, nil
+	}
+
+	return apimodels.BlockStateUnknown, nil, nil
+}
+
+func (r *Retainer) transactionStatus(blockID iotago.BlockID, block *model.Block) (bool, apimodels.TransactionState, error) {
+	// block was for sure accepted
+	if block != nil {
+		// return if block does not have a transaction
+		if _, hasTx := block.Transaction(); !hasTx {
+			return false, apimodels.TransactionStatePending, nil
+		}
+
+		if blockID.Index() <= r.finalizedSlotFunc() {
+			return true, apimodels.TransactionStateFinalized, nil
+		}
+		// check if confirmed
+		if confirmed, err := r.retainerFunc(blockID.Index()).WasTransactionConfirmed(blockID); err != nil {
+			return true, apimodels.TransactionStateFailed, ierrors.Wrap(err, "failed to get transaction confirmed state in retainer.")
+		} else if confirmed {
+			return true, apimodels.TransactionStateConfirmed, nil
+		}
+	}
+
+	if pending, err := r.retainerFunc(blockID.Index()).WasTransactionPending(blockID); err != nil {
+		return true, apimodels.TransactionStateFailed, ierrors.Wrap(err, "failed to get transaction pending state in retainer.")
```

Comment:
Mhh I wonder if we should return the failed state here, because it isn't the transaction that actually failed.